net: avoid overriding non-virtual ToString() in CService and use better naming

vasild opened the pull request:
Before this PR we had the somewhat confusing combination of methods:

`CNetAddr::ToStringIP()`
`CNetAddr::ToString()` (duplicate of the above)
`CService::ToStringIPPort()`
`CService::ToString()` (duplicate of the above, overrides a non-virtual method from `CNetAddr`)
`CService::ToStringPort()`

Avoid [overriding non-virtual methods](https://github.com/bitcoin/bitcoin/pull/25349/#issuecomment-1185226396).

"IP" stands for "Internet Protocol" and while sometimes "IP addresses" are called just "IPs", it is incorrect to call Tor or I2P addresses "IPs". Thus use "Addr" instead of "IP".

Change the above to:

`CNetAddr::ToStringAddr()`
`CService::ToStringAddrPort()`

The changes touch a lot of files, but are mostly mechanical.

vasild committed:
net: remove CNetAddr::ToString() and use ToStringAddr() instead

Both methods do the same thing, so simplify to having just one.

Further, `CService` inherits `CNetAddr` and `CService::ToString()`
overrides `CNetAddr::ToString()` but the latter is not virtual which
may be confusing. Avoid such a confusion by not having non-virtual
methods with the same names in inheritance.

vasild committed:
net: remove CService::ToString() use ToStringAddrPort() instead

Both methods do the same thing, so simplify to having just one.

`ToString()` is too generic in this case and it is unclear what it does,
given that there are similar methods:
`ToStringAddr()` (inherited from `CNetAddr`),
`ToStringPort()` and
`ToStringAddrPort()`.

MarcoFalke commented:
ToStringPort should probably be removed. The only place where it is used in the gui should probably use ToStringIPPort to properly display IPV6, no?

vasild committed:
gui: simplify OptionsDialog::updateDefaultProxyNets()

Do not create strings and compare them to check if one `addr:port`
equals another. Use `CService::operator==()` instead.

`strDefaultProxyGUI` was assigned the same value 3 times. Instead save
it in `const CService ui_proxy` at the beginning of the function.

vasild committed:
net: remove CService::ToStringPort()

It is used only internally in `CService::ToStringAddrPort()`.

jonatack commented:
Concept ACK

jonatack commented:
Light ACK, skimmed commits, debug build and unit tests. Will review more thoroughly.

luke-jr commented:
nit: "host" seems preferable to "addr"

MarcoFalke commented:
Maybe the scripted-diff can be made first to avoid repeatedly changing the same line of code several times?

vasild committed:
scripted-diff: rename ToStringIP[Port]() to ToStringAddr[Port]()

"IP" stands for "Internet Protocol".

"IP address" is sometimes shortened to just "IP" or "address".

However, Tor or I2P addresses are not "IP addresses", nor "IPs".

Thus, use "Addr" instead of "IP" for addresses that could be IP, Tor or
I2P addresses:

`CService::ToStringIPPort()` -> `CService::ToStringAddrPort()`
`CNetAddr::ToStringIP()` -> `CNetAddr::ToStringAddr()`

-BEGIN VERIFY SCRIPT-
sed -i 's/ToStringIPPort/ToStringAddrPort/g' -- $(git grep -l ToStringIPPort src)
sed -i 's/ToStringIP/ToStringAddr/g' -- $(git grep -l ToStringIP src)
-END VERIFY SCRIPT-

vasild commented:
`b9ca6d0af3...9854d3a820`: rebase due to conflicts

vasild commented:
`9854d3a820...0a798d56c2`: rebase due to conflicts

vasild commented:
`0a798d56c2...cbfa859028`: rebase due to conflicts

jonatack commented on src/qt/optionsdialog.cpp:
```diff
@@ -406,24 +406,21 @@ void OptionsDialog::updateProxyValidationState()
 {
+    CNetAddr ui_proxy_netaddr;
+    LookupHost(ui->proxyIp->text().toStdString(), ui_proxy_netaddr, /*fAllowLookup=*/false);
+    const CService ui_proxy{ui_proxy_netaddr, ui->proxyPort->text().toUShort()};
```
7f420c508968 TIL about `toUShort()` (https://doc.qt.io/qt-5/qlocale.html#toUShort) used for the first time in this codebase. It seems compatible with our minimum required version of qt 5.11.3.

vasild replied:
I wasn't aware of `toUShort()` either. IIRC the auto-completion showed it to me ;-)

jonatack reviewed (commented):
utACK cbfa85902839255a382a9f4556372ba5ceecf819 code review and debug build

Edit: might be clearest to prefix the PR title and the commit messages with `refactor:`

vasild commented on src/netaddress.cpp:
```diff
@@ -916,25 +911,17 @@ std::vector<unsigned char> CService::GetKey() const
 {
-    return strprintf("%u", port);
-}
+    const auto port_str = strprintf("%u", port);
```
There is no copy elision because [there is no copy](https://www.youtube.com/watch?v=uAXtO5dMqEI). The standard mandates that in this case there is no temporary to be copied and that the object should be [constructed directly into `port_str`](ttps://stackoverflow.com/questions/38043319/how-does-guaranteed-copy-elision-work#38043447).

vasild commented:
`cbfa859028...d4117c37f6`: rebase due to conflicts

Invalidates ACKs from @jonatack

jonatack reviewed (approved):
ACK d4117c37f6e4030af623d56c22bdb7a8c84a38c6 only change is a trivial rebase

vasild commented:
`d4117c37f6...e5d1916425`: rebase due to conflicts

Invalidates ACK from @jonatack

jonatack commented:
re-ACK e5d191642503346ba6d9fe24f0c8f4414ffa00bc only change since my last review is a trivial rebase

achow101 commented:
cc @LarryRuane @Empact

Empact commented:
Code Review ACK e5d191642503346ba6d9fe24f0c8f4414ffa00bc

vasild commented:
`e5d1916425...caeec22664`: rebase due to conflicts

Invalidates ACKs from @jonatack, @Empact

vasild commented:
`caeec22664...c9d548c91f`: rebase due to conflicts

Previously invalidated ACKs from @jonatack, @Empact

jonatack reviewed (approved):
re-ACK c9d548c91fb12fba516dee896f1f97692cfa2104 only change since my previous reviews is rebase, but as a sanity check rebased to current master and at each commit quickly re-reviewed and re-verified clean build and green unit tests

LarryRuane reviewed (commented):
ACK c9d548c91fb12fba516dee896f1f97692cfa2104
rebased to master, built, ran tests, ran `bitcoind` mainnet, code reviewed except I didn't completely understand the qt changes.
